sklearn/preprocessing/_encoders.py

```diff
@@ -0,0 +1,827 @@
+# Authors: Andreas Mueller <amueller@ais.uni-bonn.de>
+#          Joris Van den Bossche <jorisvandenbossche@gmail.com>
+# License: BSD 3 clause
+
+from __future__ import division
+
+import numbers
+import warnings
+
+import numpy as np
+from scipy import sparse
+
+from ..base import BaseEstimator, TransformerMixin
+from ..externals import six
+from ..utils import check_array
+from ..utils import deprecated
+from ..utils.fixes import _argmax
+from ..utils.validation import check_is_fitted, FLOAT_DTYPES
+from .label import LabelEncoder
+
+
+range = six.moves.range
+
+
+__all__ = [
+    'OneHotEncoder',
+    'OrdinalEncoder'
+]
+
+
+def _transform_selected(X, transform, selected="all", copy=True):
+    """Apply a transform function to portion of selected features
+
+    Parameters
+    ----------
+    X : {array-like, sparse matrix}, shape [n_samples, n_features]
+        Dense array or sparse matrix.
+
+    transform : callable
+        A callable transform(X) -> X_transformed
+
+    copy : boolean, optional
+        Copy X even if it could be avoided.
+
+    selected: "all" or array of indices or mask
+        Specify which features to apply the transform to.
+
+    Returns
+    -------
+    X : array or sparse matrix, shape=(n_samples, n_features_new)
+    """
+    X = check_array(X, accept_sparse='csc', copy=copy, dtype=FLOAT_DTYPES)
+
+    if isinstance(selected, six.string_types) and selected == "all":
+        return transform(X)
+
+    if len(selected) == 0:
+        return X
+
+    n_features = X.shape[1]
+    ind = np.arange(n_features)
+    sel = np.zeros(n_features, dtype=bool)
+    sel[np.asarray(selected)] = True
+    not_sel = np.logical_not(sel)
+    n_selected = np.sum(sel)
+
+    if n_selected == 0:
+        # No features selected.
+        return X
+    elif n_selected == n_features:
+        # All features selected.
+        return transform(X)
+    else:
+        X_sel = transform(X[:, ind[sel]])
+        X_not_sel = X[:, ind[not_sel]]
+
+        if sparse.issparse(X_sel) or sparse.issparse(X_not_sel):
+            return sparse.hstack((X_sel, X_not_sel))
+        else:
+            return np.hstack((X_sel, X_not_sel))
+
+
+class _BaseEncoder(BaseEstimator, TransformerMixin):
+    """
+    Base class for encoders that includes the code to categorize and
+    transform the input features.
+
+    """
+
+    def _fit(self, X, handle_unknown='error'):
+
+        X_temp = check_array(X, dtype=None)
```

Comment:
I know this is in CategoricalEncoder, but it might be a problem. Storing an array as strings can consume much memory.

Reply:
Maybe these days lists of lists aren't so important. I'd say in the pre-pandas days they were. They still might be to some extent, e.g. from `csv.reader`.

Reply:
That's true, however fixing this is not that trivial (you need to "sniff" the data types in the full data to check if there are strings or not), so I would leave that for a separate issue if somebody is interested in it.

Also note that this will not slowdown existing code that dates from pre-pandas times, as before this was not possible at all with strings.